From 02a1396cf7e00f2f3dae4f6dc5f2cdade9fee3c3 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:03:06 -0400
Subject: [fs] btrfs: correctly set profile flags on seqlock retry

Message-id: <1407179109-13714-300-git-send-email-sandeen@redhat.com>
Patchwork-id: 87260
O-Subject: [RHEL7.1 PATCH 299/422] Btrfs: correctly set profile flags on seqlock retry
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Filipe Manana <fdmanana@gmail.com>

commit f8213bdc89719bad895a02c62c4a85066ff76720
Author: Filipe Manana <fdmanana@gmail.com>
Date:   Thu Apr 24 15:15:29 2014 +0100

    Btrfs: correctly set profile flags on seqlock retry

    If we had to retry on the profiles seqlock (due to a concurrent write), we
    would set bits on the input flags that corresponded both to the current
    profile and to previous values of the profile.

    Signed-off-by: Filipe David Borba Manana <fdmanana@gmail.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/extent-tree.c b/fs/btrfs/extent-tree.c
index 678cb35..5590af9 100644
--- a/fs/btrfs/extent-tree.c
+++ b/fs/btrfs/extent-tree.c
@@ -3543,11 +3543,13 @@ static u64 btrfs_reduce_alloc_profile(struct btrfs_root *root, u64 flags)
  return extended_to_chunk(flags | tmp);
 }
 
-static u64 get_alloc_profile(struct btrfs_root *root, u64 flags)
+static u64 get_alloc_profile(struct btrfs_root *root, u64 orig_flags)
 {
  unsigned seq;
+ u64 flags;
 
  do {
+  flags = orig_flags;
   seq = read_seqbegin(&root->fs_info->profiles_lock);
 
   if (flags & BTRFS_BLOCK_GROUP_DATA)
-- 
1.7.1